### PR TITLE
Move to GH Actions

### PR DESCRIPTION
**Pull request by sabderemane**
Related to #344

### CODE REVIEW

**Review by ezio-melotti — COMMENTED** [review]
An early review with a few comments.

**Comment by arhadthedev** [comment]
@sabderemane Would you mind to address the reviews, please? It would be great to retire a third party Heroku service of extra maintainance, not-so-transparent deployment, non-public error logs, and a dedicated, actively waiting server loop.

**Comment by sabderemane** [comment]
I will, for sure, I just didn't have time to do so yet :)

File: .github/workflows/review_pr.yml
Comment (sabderemane) [reply .github/workflows/review_pr.yml:23]: I update the `key` but for the rest it should be ok, I use [this example](https://github.com/actions/cache/blob/main/examples.md#simple-example)

File: .github/workflows/review_pr.yml
Comment (sabderemane) [reply .github/workflows/review_pr.yml]: No, I changed

File: bedevere/backport.py
Comment (sabderemane) [reply bedevere/backport.py]: No, I only keep `event` only, changed

File: bedevere/review_pr.py
Comment (sabderemane) [reply bedevere/review_pr.py]: updated

File: .github/workflows/review_pr.yml
Comment (sabderemane) [reply .github/workflows/review_pr.yml]: Updated

**Review by sabderemane — COMMENTED** [review]
Thank you @ezio-melotti for the review 👍🏽

File: CONTRIBUTING.rst
Comment (hugovk) [inline CONTRIBUTING.rst:4]: "currently" is redundant
```suggestion
Bedevere web service is launched with GitHub Actions, which is managed by The PSF.
```

File: .github/workflows/review_pr.yml
Comment (hugovk) [reply .github/workflows/review_pr.yml:23]: `actions/setup-python` now natively includes pip caching.

Docs: https://github.com/actions/setup-python#caching-packages-dependencies

So we can remove the `uses: actions/cache@v1` step, and update:


```diff
-      - uses: actions/setup-python@v3
+      - uses: actions/setup-python@v4
         with:
           python-version: '3.x'
+          cache: pip
+          cache-dependency-path: '*requirements.txt'
```

And the same in the other workflows.

File: .github/workflows/review_pr.yml
Comment (hugovk) [inline .github/workflows/review_pr.yml:17]: Bump:
```suggestion
      - uses: actions/checkout@v3
```

File: .github/workflows/close_pr.yml
Comment (hugovk) [inline .github/workflows/close_pr.yml:17]:
```suggestion
      - uses: actions/checkout@v3
```

File: .github/workflows/close_pr.yml
Comment (hugovk) [inline .github/workflows/close_pr.yml:24]:
```suggestion
      - uses: actions/setup-python@v4
```

File: .github/workflows/backport.yml
Comment (hugovk) [inline .github/workflows/backport.yml:17]: Bump version:
```suggestion
      - uses: actions/checkout@v3
```

File: .github/workflows/backport.yml
Comment (hugovk) [inline .github/workflows/backport.yml:24]:
```suggestion
      - uses: actions/setup-python@v4
```

File: .github/workflows/filepaths.yml
Comment (hugovk) [inline .github/workflows/filepaths.yml:17]:
```suggestion
      - uses: actions/checkout@v3
```

File: .github/workflows/filepaths.yml
Comment (hugovk) [inline .github/workflows/filepaths.yml:24]:
```suggestion
      - uses: actions/setup-python@v4
```

**Review by hugovk — COMMENTED** [review]
A few workflow suggestions :)

File: .github/workflows/close_pr.yml
Comment (hugovk) [reply .github/workflows/close_pr.yml]: Dependency path: we install `dev-requirements.txt`, which includes installing `-r requirements.txt`

below I suggest using `cache-dependency-path: '*requirements.txt'`, so if either of those files change, it invalidates the cache and downloads the new things.

File: bedevere/backport.py
Comment (ezio-melotti) [inline bedevere/backport.py:38]: Why were the return values added?  ISTM that they are not used.

File: bedevere/backport.py
Comment (ezio-melotti) [inline bedevere/backport.py:82]: I think this got accidentally deleted during the merge (like the `re.IGNORECASE` above).

File: bedevere/backport.py
Comment (ezio-melotti) [inline bedevere/backport.py:92]: This should also use `not is_maintanance_branch`.
You should double check that there aren't other updates that got revert during the merge (or possibly re-do the merge).

File: bedevere/close_pr.py
Comment (ezio-melotti) [inline bedevere/close_pr.py:31]:
```suggestion
    head_label = event["pull_request"]["head"]["label"]
    base_label = event["pull_request"]["base"]["label"]
```

File: bedevere/filepaths.py
Comment (ezio-melotti) [inline bedevere/filepaths.py:18]:
```suggestion
    pull_request = event['pull_request']
```

File: bedevere/review_pr.py
Comment (ezio-melotti) [inline bedevere/review_pr.py:34]:
```suggestion
    head_label = event["pull_request"]["head"]["label"]
    base_label = event["pull_request"]["base"]["label"]
```

File: bedevere/review_pr.py
Comment (ezio-melotti) [inline bedevere/review_pr.py:42]:
```suggestion
    if (PYTHON_MAINT_BRANCH_RE.match(head_label) and
        base_label == "python:main"):
        pr = event["pull_request"]
        reviewers = [reviewer["login"] for reviewer in pr["requested_reviewers"]
        team_reviewers = [team["name"] for team in pr["requested_teams"]
        await gh.delete(f'{event["pull_request"]["url"]}/requested_reviewers',
                        data=dict(reviewers=reviewers, team_reviewers=team_reviewers))
```

**Review by ezio-melotti — CHANGES_REQUESTED** [review]
I left some more review comments.
* It seems that some of the changes applied lately got lost during a merge -- those should be double checked and fixed;
* there were some formatting problems likely caused by a misconfigured editor (trailing spaces, tabs, changed indentation)

P.S.: you can probably batch merge most of these directly from GitHub.

**Comment by sabderemane** [comment]
Sorry for the late update, I address all the issues hopefully, thank you for the review @hugovk, @ezio-melotti and @DanielNoord.  I change my configuration to avoid editor issue 🙏🏽  I will see to continue and update tests.

**Comment by arhadthedev** [comment]
From https://github.com/python/bedevere/issues/344#issuecomment-1591861785:

> One issue with using Actions is that it require maintainers to "approve" the workflow for contributors before it can be run.
The workflows done by bedevere, like checking for skip issue, skip news, applying "type" labels, should just run by itself without needing maintainer approval.
>
> So I think it would be better to convert bedevere into GitHub App instead of Actions. By converting this to GitHub App, we can resolve the issue where 2FA is needed for the bedevere-bot account.

From the following https://github.com/python/bedevere/issues/344#issuecomment-1591889387:

> To solve the issue where core devs want to be able to re-run workflows/retrigger webhook in case the web app is down, we can grant the "[GitHub App manager](https://docs.github.com/en/organizations/managing-programmatic-access-to-your-organization/adding-and-removing-github-app-managers-in-your-organization#about-github-app-managers)" access to core devs.

The problem surfaced while working on gh-568 so gh-569 was created instead.

@sabderemane Probably, your PR is a dead-end then. We need to either close it in favor of gh-569 or think on some alterations.

**Comment by sabderemane** [comment]
Closing in favor of gh-569, happy to help on other things instead :)

**Pull request closed without merging**
